KCardGrid: Part 1 - Prepare KCard + few bugfixes and a visual spacing update

MisRob opened the pull request:
Note that I worked from @AllanOXDi's https://github.com/learningequality/kolibri-design-system/pull/752 branch which hasn't been merged yet so the diff is visible in this PR. Please ignore @AllanOXDi's commits here. I will rebase as soon as #752 is merged.

## Description

This is a first part of work that was proposed and discussed in https://github.com/learningequality/kolibri-design-system/pull/748.

- Merges `KCard` and `BaseCard` to unblock several issues for the upcoming features regarding the addition of checkboxes and `KCardGrid`
- Prepares `KCard` structure and style for the aforementioned features
- Contains few bug fixes and docstrings improvements

See commit messages for details.

## References

Closes https://github.com/learningequality/kolibri-design-system/issues/773

## Changelog
<!-- [DO NOT REMOVE-USED BY GH ACTION] CHANGELOG START -->

<!--
  - Fill in the changelog item(s) below. If there are more groups of closely
    related changes, prepare more changelog items for each one of them.
    At a minimum, always separete non-breaking changes from breaking changes.
  - This needs to be pasted to CHANGELOG.md before merging a PR.
  - See changelog guidelines https://www.notion.so/learningequality/DRAFT-Changelog-Guidelines-106b6ebbdeda4ba5b3b3e7c490c5a4fe and existing
    items in CHANGELOG.md as examples
 -->

  - **Description:**  Merges `KCard` and `BaseCard`
  - **Products impact:** none
  - **Addresses:** Unblocks several issues for the upcoming features regarding the addition of checkboxes and `KCardGrid`
  - **Components:** `KCard`
  - **Breaking:** no
  - **Impacts a11y:** -
  - **Guidance:** -

  - **Description:**  Updates `KCard` internal structure and style
  - **Products impact:** This introduces temporary regressions in `KCard` related to removing its control of its height that will instead be controlled by `KCardGrid`. However, `KCard` is required to be always used within `KCardGrid` anyways so ultimately this will be of no real impact. Will be completed by `KCardGrid` soon.
  - **Addresses:**  Prepares `KCard` for the upcoming features regarding the addition of checkboxes and `KCardGrid`
  - **Components:** `KCard`
  - **Breaking:** no
  - **Impacts a11y:** -
  - **Guidance:** -
  
  - **Description:**  Fixes the thumbnail overflowing in the horizontal layout with small thumbnail and aligns this layout more closely to the designs.
  - **Products impact:** bugfix
  - **Addresses:** https://github.com/learningequality/kolibri-design-system/issues/773
  - **Components:** `KCard`
  - **Breaking:** no
  - **Impacts a11y:** -
  - **Guidance:** -
 
  - **Description:**   Fix click.stop not working on interactive elements rendered within the card via its slots.
  - **Products impact:** bugfix
  - **Addresses:** -
  - **Components:** `KCard`
  - **Breaking:** no
  - **Impacts a11y:** -
  - **Guidance:** -

 - **Description:**   Aligns padding to the designs
  - **Products impact:** Visual update
  - **Addresses:** -
  - **Components:** `KCard`
  - **Breaking:** no
  - **Impacts a11y:** -
  - **Guidance:** -


<!-- [DO NOT REMOVE-USED BY GH ACTION] CHANGELOG END -->

## Steps to test

- As this move was already discussed and agreed on, in this PR we would want to focus on catching any unexpected major regressions, particularly in the markup structure or layout. See [the playground page](https://deploy-preview-774--kolibri-design-system.netlify.app/playground/).

- Note that this introduces temporary regressions in `KCard` related to removing its control of its height (elements overflowing the card area, cards not having consistent height) that will instead be controlled by `KCardGrid`. This is deliberate step that will be completed by `KCardGrid` soon. It shouldn't cause any issues since `KCard` shouldn't be used without `KCardGrid` and consumers are instructed to link to https://github.com/learningequality/kolibri-design-system/pull/748 before all that work is transitioned to `develop`.

## Testing checklist
<!-- Complete the checklist before submitting a PR; delete anything that doesn't apply -->

- [x] Contributor has fully tested the PR manually
- [ ] ~If there are any front-end changes, before/after screenshots are included~
- [ ] ~Critical and brittle code paths are covered by unit tests~
- [x] The change is described in the changelog section above

## Reviewer guidance
<!-- Delete anything that doesn't apply so your reviewer knows what to check for -->

- [ ] Is the code clean and well-commented?
- [ ] Are there tests for this change?
- [ ] Are all UI components LTR and RTL compliant (if applicable)?
- [ ] _Add other things to check for here_

AllanOXDi committed:
fixes card display without a thumbnail

AllanOXDi committed:
lint

AllanOXDi committed:
cleans thumbnial configuration condition

AllanOXDi committed:
update changelog

AllanOXDi committed:
rebase

AllanOXDi committed:
lint

MisRob committed:
Add playground for testing

MisRob committed:
Merge BaseCard with KCard

This unblocks several issues related to the current
internal architecture - primarily gives us more
flexibility needed to achieve simple public API
for card, card grid, and card checkboxes.

MisRob committed:
Rename rootClass to layoutClass

to better reflect its purpose.

MisRob committed:
Apply card styles to an inner div

rather than outer <li>.

This prepares ground for elements
that need to be rendered within <li>
but outside the card area, such as
checkboxes.

MisRob committed:
Simplify heading styles

Previously, heading styles needed to be dynamic
so they could be passed down to BaseCard.
BaseCard is removed now so we can use much
simpler approach that also allows for removal
of the two copies of same styles in JavaScript
and CSS parts.

MisRob committed:
Simplify heading structure and style

MisRob committed:
Use consistent case

MisRob committed:
Update docstrings

- remove duplicate information
- improve generated documentation display
- improve clarity
- re-order props to related groups
- try to be conscise

MisRob committed:
Remove fixed heights

Prepares ground for card grid work
that will be responsible for controlling
card dimensions.

MisRob committed:
Remove duplicate style

`relative` is already applied
via `card-area` class.

MisRob committed:
Update thumbnail area size

in the horizontal layout with small thumbnail.

Fixes the thumbnail overflowing
and aligns this layout more closely
to the designs.

MisRob committed:
Reset list item styles

MisRob committed:
Improve tests clarity

AllanOXDi commented:
Hi @MisRob thanks for the wonderful work here.  I am leaving few questions or comments here.

AllanOXDi reviewed (approved):
I am not sure if we could keep cards to uniform   sizes  of the card in this PR regardless whether the slots/ thumbnail is provided. I feel like something changed along the way

<img width="1414" alt="Screenshot 2024-09-11 at 17 27 52" src="https://github.com/user-attachments/assets/e746defe-359b-47d5-ad38-7620901a7bed">
<img width="464" alt="Screenshot 2024-09-11 at 17 32 07" src="https://github.com/user-attachments/assets/5c1db552-6ed5-4866-bb43-00799bf20b98">

MisRob commented:
Hi @AllanOXDi, ah yes that's right. Thank you. All these issues are addressed in the PR that comes after this one. All of them are the temporary regressions I meant in the "Steps to test" - I will be more specific next time. This is caused by removing the height control from `KCard` and transitioning it to `KCardGrid` that will control the row height :)  You can preview how it will eventually look like in https://github.com/learningequality/kolibri-design-system/pull/748 playground - I'm just breaking it down to smaller PRs.

MisRob committed:
Fix click.stop not working on interactive elements

rendered within the card via its slots.

MisRob committed:
Align spaces to the designs

MisRob commented:
I am adding
- A fix for `@click.stop` on interactive elements not working (https://github.com/learningequality/kolibri/pull/12418#issuecomment-2341380628) 
- Update of spaces to 16px as agreed with designers ([Slack thread](https://learningequality.slack.com/archives/C0LKG14NL/p1725960788683929))

MisRob committed:
Merge branch 'develop' into k-card-grid-1

MisRob committed:
Fix tests

'mouseup' was replaced by 'click' event
in the previous work so tests
need to reflect it

AlexVelezLl commented on lib/KCard/index.vue:
```diff
@@ -1,20 +1,59 @@
+        :style="{ color: $themeTokens.text }"
+      >
+        <!--
+          Prevent router-link click event by setting empty event=""
```
I know this was moved from BaseCard, but I dont understand well what empty event this refers to, perhaps the comment is about why the router-link has tabindex="-1"?

MisRob replied:
Oh I think previously there was `event=""` applied to `router-link` and this comment was about that. I must have lost it during the refactor. I will investigate if it's still needed. Good catch!

MisRob replied:
Yup, I need to return it to keep the feature for making card's text selectable. Done, also tried to improve the comment.

AlexVelezLl commented on lib/KCard/index.vue:
```diff
@@ -109,38 +146,54 @@
+      },
+      /**
+       * Specifies the number of lines allowed
+       * for the titlebefore truncation occurs.
```
Just a little typo here
```suggestion
       * for the title before truncation occurs.
```

And I think is better to specify that is the _max_ number of lines before truncate it. I think right now it can be confused with "number of lines we will be reserving for every title" (so, if the title just have 1 line, there will be an empty line).

MisRob replied:
I changed it to `Truncates title to this number of lines.`

AlexVelezLl commented on lib/KCard/index.vue:
```diff
@@ -149,84 +202,64 @@
-       * @default false
+       * If `aboveTitle` slot is empty, this controls whether
+       * its space is preserved or not. Typically used to achieve
+       * consistent vertical alignment of information when there
```
Loved this description <3

MisRob replied:
Thanks :)

AlexVelezLl reviewed (approved):
This looks good to me! I think the merge of BaseCard and KCard has had a good and cleaner result. Just left 2 super minor things in docs wording now that you already did some changes there :sweat_smile:. Thank you!

MisRob commented:
> . Just left 2 super minor things in docs wording now that you already did some changes there 😅.

No worries @AlexVelezLl, I welcome that. Will look into it :) Thank you.

MisRob committed:
Fix the selectable text feature

MisRob committed:
Simplify docstring and make more clear
